Search (ala SearchView) doesn't work with UUID pk's(Non Int pk's in general)

* [x] Tested with the latest Haystack release
* [x] Tested with the current Haystack master branch

## Expected behaviour
You could use UUID pk's for models in django.  Setup a SearchView and it all works.

## Actual behaviour
You can see the number of results with sqs.all().count() in for example your SearchForm Class.  But the moment you try to instantiate and get all the results...it evaporates.  No results available when you try to iterate over the list.

## Steps to reproduce the behaviour

1.Use UUID's as primary keys on django models that are indexed
2.Create a SearchView and SearchForm that use the Model.
3.Try to use the sqs in the html template through object_list OR a paginated result set (page_obj.object_list)

## Configuration

* Operating system version:Ubuntu 16.04
* Search engine version:6.5
* Python version:3.5.2
* Django version:1.11.4
* Haystack version:master branch and earlier

## Root Cause Analysis
Still in process but it looks like post_process_results of query.py:SearchQuerySet tries to force an int() conversion on the pk causing a problem...or it's involved somehow.  resulting in every result being discarded as an ignored result silently

Still working out the exact nature.

This took me forever to figure out.  I thought I had it working before...I don't know what changed to make it stop.

Trying to get tests fully passing.  Ran into one weird problem locally where solr tests would fail on 
```
FAIL: test_related_load_all_queryset (test_haystack.solr_tests.test_solr_backend.LiveSolrSearchQuerySetTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dev/django-haystack/test_haystack/solr_tests/test_solr_backend.py", line 1061, in test_related_load_all_queryset
    self.assertEqual([obj.object.id for obj in sqs], list(range(11, 24)))
AssertionError: Lists differ: [] != [11, 12, 13, 14, 15, 16, 17, 1...

Second list contains 13 additional elements.
First extra element 0:
11

- []
+ [11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]
```

I don't think changes should really have changed those results.  Still testing etc.

--- a/haystack/query.py
+++ b/haystack/query.py
@@ -172,11 +172,14 @@ class SearchQuerySet(object):
 
         for result in results:
             if self._load_all:
-                # We have to deal with integer keys being cast from strings
+                
                 model_objects = loaded_objects.get(result.model, {})
+                # Try to coerce a primary key object that matches the models pk
+                # We have to deal with semi-arbitrary keys being cast from strings (UUID, int, etc)
                 if result.pk not in model_objects:
                     try:
-                        result.pk = int(result.pk)
+                        result_klass = type(next(iter(model_objects)))
+                        result.pk = result_klass(result.pk)
                     except ValueError:
                         pass
                 try:

--- a/haystack/query.py
+++ b/haystack/query.py
@@ -172,26 +172,32 @@ class SearchQuerySet(object):
 
         for result in results:
             if self._load_all:
-                
+
                 model_objects = loaded_objects.get(result.model, {})
                 # Try to coerce a primary key object that matches the models pk
                 # We have to deal with semi-arbitrary keys being cast from strings (UUID, int, etc)
-                if result.pk not in model_objects:
-                    try:
-                        result_klass = type(next(iter(model_objects)))
-                        result.pk = result_klass(result.pk)
-                    except ValueError:
-                        pass
-                try:
-                    result._object = model_objects[result.pk]
-                except KeyError:
-                    # The object was either deleted since we indexed or should
-                    # be ignored; fail silently.
-                    self._ignored_result_count += 1
+                if model_objects:
+                    result_klass = type(next(iter(model_objects)))
+                    result.pk = result_klass(result.pk)
 
-                    # avoid an unfilled None at the end of the result cache
-                    self._result_cache.pop()
-                    continue
+                    try:
+                        result._object = model_objects[result.pk]
+                    except KeyError:
+                        # The object was either deleted since we indexed or should
+                        # be ignored for other reasons such as an overriden 'load_all_queryset';
+                        # fail silently.
+                        self._ignored_result_count += 1
+
+                        # avoid an unfilled None at the end of the result cache
+                        self._result_cache.pop()
+                        continue
+                else:
+                    # No objects were returned -- possible due to SQS nesting such as
+                    # XYZ.objects.filter(id__gt=10) where the amount ignored are
+                    # exactly equal to the ITERATOR_LOAD_PER_QUERY
+                    del self._result_cache[:len(results)]
+                    self._ignored_result_count += len(results)
+                    break
 
             to_cache.append(result)
 

--- a/test_haystack/core/fixtures/base_data.json
+++ b/test_haystack/core/fixtures/base_data.json
@@ -71,4 +71,18 @@
       "deleted": true
     }
   }
+  {
+    "pk": "53554c58-7051-4350-bcc9-dad75eb248a9",
+    "model": "core.uuidmockmodel",
+    "fields": {
+      "characteristics": "some text that was indexed",
+    }
+  }
+  {
+    "pk": "77554c58-7051-4350-bcc9-dad75eb24888",
+    "model": "core.uuidmockmodel",
+    "fields": {
+      "characteristics": "more text that was indexed",
+    }
+  }
 ]

--- a/test_haystack/core/fixtures/base_data.json
+++ b/test_haystack/core/fixtures/base_data.json
@@ -70,19 +70,19 @@
       "author": "sam2",
       "deleted": true
     }
-  }
+  },
   {
     "pk": "53554c58-7051-4350-bcc9-dad75eb248a9",
     "model": "core.uuidmockmodel",
     "fields": {
-      "characteristics": "some text that was indexed",
+      "characteristics": "some text that was indexed"
     }
-  }
+  },
   {
     "pk": "77554c58-7051-4350-bcc9-dad75eb24888",
     "model": "core.uuidmockmodel",
     "fields": {
-      "characteristics": "more text that was indexed",
+      "characteristics": "more text that was indexed"
     }
   }
 ]

--- a/test_haystack/core/models.py
+++ b/test_haystack/core/models.py
@@ -25,7 +25,7 @@ class MockModel(models.Model):
     def hello(self):
         return 'World!'
 
-class MockUUIDModel(models.Model):
+class UUIDMockModel(models.Model):
     id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
     characteristics = models.TextField()
 

--- a/test_haystack/mocks.py
+++ b/test_haystack/mocks.py
@@ -108,6 +108,11 @@ class CharPKMockSearchBackend(MockSearchBackend):
     mock_search_results = [MockSearchResult('core', 'CharPKMockModel', 'sometext', 0.5),
                            MockSearchResult('core', 'CharPKMockModel', '1234', 0.3)]
 
+class UUIDMockSearchBackend(MockSearchBackend):
+    model_name = 'uuidmockmodel'
+    mock_search_results = [MockSearchResult('core', 'UUIDMockModel', '53554c58-7051-4350-bcc9-dad75eb248a9', 0.5),
+                           MockSearchResult('core', 'UUIDMockModel', '77554c58-7051-4350-bcc9-dad75eb24888', 0.5)]
+
 
 class ReadQuerySetMockSearchBackend(MockSearchBackend):
     model_name = 'afifthmockmodel'

--- a/test_haystack/test_query.py
+++ b/test_haystack/test_query.py
@@ -6,7 +6,7 @@ import unittest
 
 from django.test import TestCase
 from django.test.utils import override_settings
-from test_haystack.core.models import AnotherMockModel, CharPKMockModel, MockModel
+from test_haystack.core.models import AnotherMockModel, CharPKMockModel, MockModel, UUIDMockModel
 
 from haystack import connections, indexes, reset_search_queries
 from haystack.backends import SQ, BaseSearchQuery
@@ -16,7 +16,7 @@ from haystack.query import EmptySearchQuerySet, SearchQuerySet, ValuesListSearch
 from haystack.utils.loading import UnifiedIndex
 
 from .mocks import (MOCK_SEARCH_RESULTS, CharPKMockSearchBackend, MockSearchBackend, MockSearchQuery,
-                    ReadQuerySetMockSearchBackend)
+                    ReadQuerySetMockSearchBackend, UUIDMockSearchBackend)
 from .test_indexes import GhettoAFifthMockModelSearchIndex, TextReadQuerySetTestSearchIndex
 from .test_views import BasicAnotherMockModelSearchIndex, BasicMockModelSearchIndex
 
@@ -323,6 +323,11 @@ class CharPKMockModelSearchIndex(indexes.SearchIndex, indexes.Indexable):
     def get_model(self):
         return CharPKMockModel
 
+class SimpleMockUUIDModelIndex(indexes.SearchIndex, indexes.Indexable):
+    text = indexes.CharField(document=True, model_attr="characteristics")
+
+    def get_model(self):
+        return UUIDMockModel
 
 @override_settings(DEBUG=True)
 class SearchQuerySetTestCase(TestCase):
@@ -423,7 +428,6 @@ class SearchQuerySetTestCase(TestCase):
         results = self.msqs.all()
         loaded = [result.pk for result in results._manual_iter()]
         self.assertEqual(loaded, [u'53554c58-7051-4350-bcc9-dad75eb248a9', u'77554c58-7051-4350-bcc9-dad75eb24888'])
-        self.assertEqual(len(connections['default'].queries), 1)
 
         connections['default']._index = old_ui
 

--- a/test_haystack/test_query.py
+++ b/test_haystack/test_query.py
@@ -336,7 +336,8 @@ class SearchQuerySetTestCase(TestCase):
         self.ui = UnifiedIndex()
         self.bmmsi = BasicMockModelSearchIndex()
         self.cpkmmsi = CharPKMockModelSearchIndex()
-        self.ui.build(indexes=[self.bmmsi, self.cpkmmsi])
+        self.uuidmmsi = SimpleMockUUIDModelIndex()
+        self.ui.build(indexes=[self.bmmsi, self.cpkmmsi, self.uuidmmsi])
         connections['default']._index = self.ui
 
         # Update the "index".
@@ -402,6 +403,8 @@ class SearchQuerySetTestCase(TestCase):
         # Test to ensure we properly fill the cache, even if we get fewer
         # results back (not a handled model) than the hit count indicates.
         # This will hang indefinitely if broken.
+
+        # CharPK testing
         old_ui = self.ui
         self.ui.build(indexes=[self.cpkmmsi])
         connections['default']._index = self.ui
@@ -410,6 +413,16 @@ class SearchQuerySetTestCase(TestCase):
         results = self.msqs.all()
         loaded = [result.pk for result in results._manual_iter()]
         self.assertEqual(loaded, [u'sometext', u'1234'])
+        self.assertEqual(len(connections['default'].queries), 1)
+
+        #UUID testing
+        self.ui.build(indexes=[self.uuidmmsi])
+        connections['default']._index = self.ui
+        self.uuidmmsi.update()
+
+        results = self.msqs.all()
+        loaded = [result.pk for result in results._manual_iter()]
+        self.assertEqual(loaded, [u'53554c58-7051-4350-bcc9-dad75eb248a9', u'77554c58-7051-4350-bcc9-dad75eb24888'])
         self.assertEqual(len(connections['default'].queries), 1)
 
         connections['default']._index = old_ui
@@ -519,6 +532,14 @@ class SearchQuerySetTestCase(TestCase):
         # Models with character primary keys.
         sqs = SearchQuerySet()
         sqs.query.backend = CharPKMockSearchBackend('charpk')
+        results = sqs.load_all().all()
+        self.assertEqual(len(results._result_cache), 0)
+        results._fill_cache(0, 2)
+        self.assertEqual(len([result for result in results._result_cache if result is not None]), 2)
+
+        # Models with uuid primary keys.
+        sqs = SearchQuerySet()
+        sqs.query.backend = UUIDMockSearchBackend('uuid')
         results = sqs.load_all().all()
         self.assertEqual(len(results._result_cache), 0)
         results._fill_cache(0, 2)

